### PR TITLE
feat: create push/unregister lambda

### PR DESCRIPTION
**Pull request by alexruzenhack**
### Acceptance Criteria
- create push/unregister lambda

### Security Checklist
- [ ] Make sure you do not include new dependencies in the project unless strictly necessary and do not include dev-dependencies as production ones. More dependencies increase the possibility of one of them being hijacked and affecting us.

### CODE REVIEW

File: src/db/index.ts [hunk 1]
```diff
@@ -2753,3 +2753,24 @@ export const updatePushDevice = async (
     [input.enablePush, input.enableShowAmounts, input.deviceId, input.walletId],
   );
 };
+
+/**
+ * Unregister push device for a given wallet.
+ *
+ * @param mysql - Database connection
+ * @param input - Input of push device register
```
Comment (pedroferreira1) [inline src/db/index.ts]: We don't have this parameter

File: tests/db.test.ts [hunk 2]
```diff
@@ -2299,3 +2300,33 @@ test('removeAllPushDeviceByDeviceId', async () => {
...
+
+  await expect(checkPushDevicesTable(mysql, 1, {
+    walletId,
+    deviceId,
+    pushProvider,
+    enablePush,
+    enableShowAmounts,
+  })).resolves.toBe(true);
+
+  await unregisterPushDevice(mysql, deviceId, walletId);
```
Comment (pedroferreira1) [inline tests/db.test.ts]: Maybe add another `checkPushDevicesTable` and expect it to be `false` to validate it has unregistered the device?

File: src/api/pushUnregister.ts [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+  const body: PushDelete = value;
+  await unregisterPushDevice(mysql, body.deviceId, walletId);
+
+  return {
+    statusCode: 200,
+    body: JSON.stringify({ success: true }),
+  };
+}))
+  .use(cors())
+  .use(warmupMiddleware());
```
Comment (tuliomir) [inline src/api/pushUnregister.ts]: Same question about the walletWarmer that is defined on the `serverless.yml` as disabled, but added here to the handler.
Should we have this?

File: tests/db.test.ts [hunk 4]
```diff
@@ -2299,3 +2300,33 @@ test('removeAllPushDeviceByDeviceId', async () => {
   await removeAllPushDeviceByDeviceId(mysql, deviceId);
   await expect(checkPushDevicesTable(mysql, 0)).resolves.toBe(true);
 });
+
+test('unregisterPushDevice', async () => {
```
Comment (tuliomir) [inline tests/db.test.ts]: I feel we should also have tests for:
- the behavior when there is no device to unregister
- assert that, when multiple devices are registered for the same wallet, only the correct device is removed

File: tests/pushUnregister.test.ts [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+beforeEach(async () => {
+  await cleanDatabase(mysql);
+});
+
+afterAll(async () => {
+  await closeDbConnection(mysql);
+});
+
+test('update push device given a wallet', async () => {
```
Comment (tuliomir) [inline tests/pushUnregister.test.ts]:
```suggestion
test('unregister push device for a given wallet', async () => {
```

File: tests/pushUnregister.test.ts [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+    pushProvider,
+    enablePush: true,
+    enableShowAmounts,
+  })).resolves.toBe(true);
+
+  const event = makeGatewayEventWithAuthorizer(walletId, null, {
+    deviceId,
+  });
+
+  const result = await unregister(event, null, null) as APIGatewayProxyResult;
```
Comment (tuliomir) [inline tests/pushUnregister.test.ts:65]: While not necessary, I feel we should have a test that tries to delete a device that was never registered, to ensure its result will be a success.

**Review by tuliomir — APPROVED** [review]

File: src/api/pushUnregister.ts [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+  static readonly bodySchema = Joi.object({
+    deviceId: Joi.string().max(256).required(),
+  });
+
+  static validate(payload: unknown): { value: PushDelete, error: ValidationError } {
+    const { value, error } = PushUpdateUnregisterValidator.bodySchema.validate(payload, {
+      abortEarly: false, // We want it to return all the errors not only the first
+      convert: true, // We need to convert as parameters are sent on the QueryString
+    });
+    return { value: { enablePush: false, enableShowAmounts: false, ...value }, error };
```
Comment (pedroferreira1) [inline src/api/pushUnregister.ts]: What are those `enablePush: false, enableShowAmounts: false` values? Do we use them here?

Comment (andreabadesso) [reply src/api/pushUnregister.ts]: Apparently they are default values

@alexruzenhack You can use `Joi` to set default values, like `.default(true)`

File: src/db/index.ts [hunk 8]
```diff
@@ -2750,6 +2750,26 @@ export const updatePushDevice = async (
...
+ */
+export const unregisterPushDevice = async (
+  mysql: ServerlessMysql,
+  deviceId: string,
+  walletId: string,
+) : Promise<void> => {
+  await mysql.query(
+    `
+    DELETE FROM \`push_devices\`
+    WHERE device_id = ? AND wallet_id = ?`,
```
Comment (andreabadesso) [inline src/db/index.ts]:
```suggestion
    DELETE
      FROM \`push_devices\`
     WHERE \`device_id\` = ?
       AND \`wallet_id\` = ?`,
```

**Review by andreabadesso — CHANGES_REQUESTED** [review]

**Review by pedroferreira1 — APPROVED** [review]

**Review by andreabadesso — APPROVED** [review]

**Review by andreabadesso — APPROVED** [review]

**Pull request merged**